### PR TITLE
implemented round() function

### PR DESCRIPTION
**Pull request by VidhyaVarshanyJS**
I have made changes the files
image
Kindly review my implementation.

### CODE REVIEW

File: temporian/core/event_set_ops.py [hunk 1]
```diff
@@ -2895,6 +2927,7 @@ def log(self: EventSetOrNode) -> EventSetOrNode:
 
         return log(self)
 
+
```
Comment (ianspektor) [inline temporian/core/event_set_ops.py:2930]: extra newline here - check that your black formatting is correctly configured to run on save and using the configuration in pyproject.toml (this is already configured in `.vscode/settings.json` and should be working unless you've overriden some of the settings)

File: temporian/core/event_set_ops.py [hunk 2]
```diff
@@ -1501,6 +1501,38 @@ def abs(
...
+            ...
+
+            ```
+
+        Returns:
+            EventSet with rounded feature values.
+        """
+        from temporian.core.operators.unary import round
+
+        return round(self)
```
Comment (ianspektor) [inline temporian/core/event_set_ops.py:1533]: 👍🏼

File: temporian/core/operators/unary.py [hunk 3]
```diff
@@ -182,11 +182,30 @@ def get_output_dtype(cls, feature_dtype: DType) -> DType:
...
+    @classmethod
+    def allowed_dtypes(cls) -> List[DType]:
+        return [
+            DType.INT32,
+            DType.INT64,
+        ]
+
+    @classmethod
+    def get_output_dtype(cls, feature_dtype: DType) -> DType:
+        return feature_dtype
```
Comment (ianspektor) [inline temporian/core/operators/unary.py:199]: These look off:
- `allowed_dtypes` specifies the types that the operator can consume - should be DType.FLOAT32 and DType.FLOAT64
- `get_output_dtype` returns the output dtype, given the input dtype - in this case it should return the same type as received (I said it should return ints in the issue, edited it, needs to return floats because they have a much larger range than ints)

File: temporian/core/operators/unary.py [hunk 4]
```diff
@@ -182,11 +182,30 @@ def get_output_dtype(cls, feature_dtype: DType) -> DType:
...
+
+
 operator_lib.register_operator(InvertOperator)
 operator_lib.register_operator(IsNanOperator)
 operator_lib.register_operator(NotNanOperator)
 operator_lib.register_operator(AbsOperator)
 operator_lib.register_operator(LogOperator)
+operator_lib.register_operator(RoundOperator)
+
```
Comment (ianspektor) [inline temporian/core/operators/unary.py:208]: extra newline here too

**Review by ianspektor — COMMENTED** [review]
This is a great start! 🚀 

Missing:
- Tests
    - Be thorough
    - Test edge cases
    - Test passing an EventSet with a single feature
    - Test passing an EventSet with several features
    - Test passing an EventSet with features of non-accepted types (e.g. string, ints)
    - Test passing an EventSet with some float32 and some float64 features, ensure output types are the same as input ones
- Adding the new operation to the docs

See last two points in https://temporian.readthedocs.io/en/latest/contributing/#developing-a-new-operator for guidance and examples.

Let's keep this PR for the round() function only, and open a new one for floor() and ceil() once done.

Thanks!

**Comment by VidhyaVarshanyJS** [comment]
> This is a great start! 🚀
> 
> Missing:
> 
> * Tests
>   
>   * Be thorough
>   * Test edge cases
>   * Test passing an EventSet with a single feature
>   * Test passing an EventSet with several features
>   * Test passing an EventSet with features of non-accepted types (e.g. string, ints)
>   * Test passing an EventSet with some float32 and some float64 features, ensure output types are the same as input ones
> * Adding the new operation to the docs
> 
> See last two points in https://temporian.readthedocs.io/en/latest/contributing/#developing-a-new-operator for guidance and examples.
> 
> Let's keep this PR for the round() function only, and open a new one for floor() and ceil() once done.
> 
> Thanks!

Thanks for your time!
Shall  I create a separate `test_round.py`  or implement test cases within the `test_unary.py` itself?

**Comment by ianspektor** [comment]
> Shall I create a separate test_round.py or implement test cases within the test_unary.py itself?

Let's do `test_unary.py`, since they're implemented in the same file too.

**Comment by VidhyaVarshanyJS** [comment]
Hi..
I have added the test cases for the round() in `test_unary.py` for the all the above combinations that is mentioned.
Kindly review my changes.
Thanks

File: temporian/core/operators/test/test_unary.py [hunk 5]
```diff
@@ -99,6 +99,53 @@ def test_correct_notnan(self) -> None:
...
+        assertOperatorResult(self, evset.round(), expected)
+        assertOperatorResult(self, round(evset), expected)  # __round__ magic
+
+    def test_round_non_accepted_types(self):
+        evset = event_set(
+            timestamps=[1, 2],
+            features={"a": ["10.5", 11.7], "b": [1, 2]},
+        )
+        with self.assertRaises(TypeError):
+            _ = evset
```
Comment (ianspektor) [inline temporian/core/operators/test/test_unary.py:134]: missing calling `round()` on the evset here - this shouldn't be raising anything? did you manage to run the tests locally to ensure they pass?

File: temporian/core/operators/test/test_unary.py [hunk 6]
```diff
@@ -99,6 +99,53 @@ def test_correct_notnan(self) -> None:
...
+        assertOperatorResult(self, round(evset), expected)  # __round__ magic
+
+    def test_round_multiple_features(self):
+        evset = event_set(
+            timestamps=[1, 2],
+            features={"a": [10.5, 11.7], "b": [1.2, 2.9]},
+        )
+        expected = event_set(
+            timestamps=[1, 2],
+            features={"a": [11, 12], "b": [1, 3]},
```
Comment (ianspektor) [inline temporian/core/operators/test/test_unary.py:122]: I don't think these will pass, since a list of ints will yield an eventset with int features, which won't be equal to the float results of round(). Ensure your new tests are passing by running `bazel test //temporian/core/operators/test:test_unary --config=macos --test_output=errors` or `bazel test //temporian/core/operators/test:test_unary --config=linux --test_output=errors` depending on your OS

File: temporian/core/operators/test/test_unary.py [hunk 7]
```diff
@@ -99,6 +99,53 @@ def test_correct_notnan(self) -> None:
...
+    def test_round_float32_and_float64_features(self):
+        evset = event_set(
+            timestamps=[1, 2],
+            features={"a": [10.5, 11.7], "b": [1.2, 2.9]},
+        )
+        expected = event_set(
+            timestamps=[1, 2],
+            features={"a": [11.0, 12.0], "b": [1.0, 3.0]},
+            same_sampling_as=evset,
+        )
```
Comment (ianspektor) [inline temporian/core/operators/test/test_unary.py:145]: which of these are being defined as f32 and which as f64? see the `f64` and `f32` methods in `temporian/test/utils.py` to explicitly create an eventset with the desired feature types

File: temporian/core/operators/unary.py [hunk 8]
```diff
@@ -182,11 +182,31 @@ def get_output_dtype(cls, feature_dtype: DType) -> DType:
...
+    def op_key_definition(cls) -> str:
+        return "ROUND"
+
+    @classmethod
+    def allowed_dtypes(cls) -> List[DType]:
+        return [
+            DType.FLOAT32,
+            DType.FLOAT64,
+            DType.INT32,
+            DType.INT64,
```
Comment (ianspektor) [inline temporian/core/operators/unary.py:196]: ints shouldn't be allowed

**Review by ianspektor — APPROVED** [review]

**Review by ianspektor — CHANGES_REQUESTED** [review]
Good work so far, please run tests locally to make sure they pass before submitting

File: temporian/core/event_set_ops.py [hunk 9]
```diff
@@ -1501,6 +1501,38 @@ def abs(
...
+    def __round__(self):
+        from temporian.core.operators.unary import round
+
+        return round(input=self)
+
+    def round(
+        self: EventSetOrNode,
+    ) -> EventSetOrNode:
+        """Rounds the values of an [`EventSet`][temporian.EventSet]'s features to the nearest integer.
+
```
Comment (ianspektor) [inline temporian/core/event_set_ops.py:1513]: Add another line specifying that only float types are allowed, and the output type will always be the same as the input's

File: temporian/core/event_set_ops.py [hunk 10]
```diff
@@ -1501,6 +1501,38 @@ def abs(
 
         return abs(self)
 
+    def __round__(self):
+        from temporian.core.operators.unary import round
+
+        return round(input=self)
+
+    def round(
```
Comment (ianspektor) [inline temporian/core/event_set_ops.py:1509]: Missing adding the new op to the documentation: 
- missing `docs/src/reference/temporian/operators/unary/round.md` file (`docs/src/reference/temporian/operators/unary/abs.md` for example)
- missing adding the op to `docs/src/reference/index.md` (see `abs` definition for example, `round` should be added as a new line in the same table, respecting alphabetical order, making sure the prettier extension is working so that the markdown formatting is correct)

you can run `mkdocs serve -f docs/mkdocs.yml` to preview your changes to the docs locally

**Review by javiber — CHANGES_REQUESTED** [review]
Looking good, There are some leftovers from the merge on `index.md`
Black should take care of the incorrect indentations and unnecessary empty lines. If you have any issue running black reach out to me on discord
